include/cuda/memory_resource

```diff
@@ -0,0 +1,809 @@
+
+#ifndef _CUDA_MEMORY_RESOURCE
+#define _CUDA_MEMORY_RESOURCE
+
+#include <memory>
+#include "std/cstddef"
+#include "std/detail/__config"
+#include "std/detail/__pragma_push"
+#include "std/type_traits"
+#include "std/utility"
+#include "std/version"
+#include "stream_view"
+
+#if _LIBCUDACXX_STD_VER > 14
+#include <memory_resource>
+#endif // _LIBCUDACXX_STD_VER > 14
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+/*!
+ * \brief Groups the tag types denoting the kind of memory of an allocation.
+ *
+ * Memory allocation kind determines where memory can be accessed and the
+ * performance characteristics of accesses.
+ *
+ * This is not a closed set, the user code can define custom memory kinds.
+ */
+namespace memory_kind {
+  /*!
+   * \brief Ordinary host memory
+   */
+  struct host;
+
+  /*!
+   * \brief Device memory, as allocated by cudaMalloc.
+   */
+  struct device;
+
+  /*!
+   * \brief Device-accessible host memory.
+   */
+  struct pinned;
+
+  /*!
+   * \brief Virtual memory that is automatically migrated between the host and devices.
+   */
+  struct managed;
+};
+
+/*!
+ * \brief Tag type for the default context of `memory_resource`.
+ *
+ * Default context in which storage may be used immediately on any thread or any
+ * CUDA stream without synchronization.
+ */
+struct any_context{};
+
+namespace detail {
+template <typename _Context>
+class __get_context_impl {
+protected:
+  virtual _Context do_get_context() const noexcept = 0;
+};
+
+template<>
+class __get_context_impl<any_context> {
+protected:
+  virtual any_context do_get_context() const noexcept { return any_context{}; }
+};
+
+}  // namespace detail
+
+template <typename _ResourcePointer, typename... _Properties>
+class basic_resource_view;
+
+/*!
+ * \brief Groups the tag types denoting the execution environment in which the memory can be accessed
+ *
+ * This is not a closed set, the user code can define custom accessibility.
+ */
+namespace memory_access {
+    struct host;
+    struct device;
+}
+
+/*!
+ * \brief A memory property tag type indicating that the memory can be oversubscribed.
+ *
+ * Oversubscribable memory doesn't need to have backing physical storage at all times.
+ */
+struct oversubscribable;
+
+/*!
+ * \brief A memory property tag type indicating that the memory has a backing physical
+ *        storage in the target location at all times.
+ */
+struct resident;
+
+/*!
+ * \brief Groups the tag types that denote the actual location of the physical storage
+ *
+ * Memory kinds which can be migrated between locations can define multiple locations.
+ */
+namespace memory_location {
+    /*!
+     * \brief A memory property tag type indicating that the memory is located on a device
+     */
+    struct device;
+
+    /*!
+    * \brief A memory property tag type indicating that the memory is located in the host memory
+    */
+    struct host;
+}
+
+
+class memory_resource_base {
```

Comment:
`memory_resource_base` should probably be in `detail::`. I don't believe we expect users to be interacting with this class directly.

Reply:
Will fix.